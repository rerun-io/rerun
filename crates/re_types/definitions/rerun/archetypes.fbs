include "./archetypes/annotation_context.fbs";
include "./archetypes/arrows3d.fbs";
include "./archetypes/asset3d.fbs";
include "./archetypes/bar_chart.fbs";
include "./archetypes/boxes2d.fbs";
include "./archetypes/boxes3d.fbs";
include "./archetypes/clear.fbs";
include "./archetypes/depth_image.fbs";
include "./archetypes/disconnected_space.fbs";
include "./archetypes/image.fbs";
include "./archetypes/line_strips2d.fbs";
include "./archetypes/line_strips3d.fbs";
include "./archetypes/mesh3d.fbs";
include "./archetypes/pinhole.fbs";
include "./archetypes/points2d.fbs";
include "./archetypes/points3d.fbs";
include "./archetypes/segmentation_image.fbs";
include "./archetypes/tensor.fbs";
include "./archetypes/text_document.fbs";
include "./archetypes/text_log.fbs";
include "./archetypes/time_series_scalar.fbs";
include "./archetypes/transform3d.fbs";
include "./archetypes/view_coordinates.fbs";

include "./testing/archetypes/fuzzy.fbs";

// NOTE: The current file is the entrypoint for codegen, and so we need to import
// the blueprint types here otherwise nothing will.
include "./blueprint.fbs";

namespace rerun.archetypes;
