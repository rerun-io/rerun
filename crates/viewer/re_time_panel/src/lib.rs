//! Rerun Time Panel
//!
//! This crate provides a panel that shows all entities in the store and allows control of time and
//! timelines, as well as all necessary ui elements that make it up.

mod data_density_graph;
mod paint_ticks;
mod recursive_chunks_per_timeline_subscriber;
mod streams_tree_data;
mod time_axis;
mod time_control_ui;
mod time_panel;
mod time_ranges_ui;
mod time_selection_ui;

pub use time_panel::TimePanel;

#[cfg(feature = "testing")]
pub use {streams_tree_data::StreamsTreeData, time_panel::TimePanelSource};

#[doc(hidden)]
pub mod __bench {
    pub use crate::data_density_graph::*;
    pub use crate::time_panel::TimePanelItem;
    pub use crate::time_ranges_ui::TimeRangesUi;
}
