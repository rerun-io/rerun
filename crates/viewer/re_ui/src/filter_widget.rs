use egui::{Color32, NumExt as _, Widget as _};
use itertools::Either;
use rand::random;

use crate::{list_item, UiExt as _};

/// State for the filter widget when it is toggled on.
#[derive(Debug, Clone)]
struct InnerState {
    /// The filter query string.
    filter_query: String,

    /// This ID is recreated every time the filter is toggled and tracks the current filtering
    /// session.
    ///
    /// This can be useful for client code to store session-specific state (e.g., the state of tree
    /// collapsed-ness).
    session_id: egui::Id,
}

impl Default for InnerState {
    fn default() -> Self {
        Self {
            filter_query: String::new(),

            // create a new session id each time the filter is toggled
            session_id: egui::Id::new(random::<u64>()),
        }
    }
}

/// State and UI for the filter widget.
///
/// The filter widget is designed as a toggle between a title widget and the filter text field.
/// [`Self`] is responsible for storing the widget state as well as the query text typed by the
/// user. [`FilterMatcher`] performs the actual filtering.
#[derive(Debug, Clone, Default)]
pub struct FilterState {
    /// The current state of the filter widget.
    ///
    /// This is `None` when the filter is not active.
    inner_state: Option<InnerState>,

    /// Should the text field be focused?
    ///
    /// Set to `true` upon clicking on the search button.
    request_focus: bool,
}

impl FilterState {
    /// Return the filter if any.
    ///
    /// The widget must be enabled _and_ the filter string must not be empty.
    pub fn query(&self) -> Option<&str> {
        self.inner_state
            .as_ref()
            .map(|state| state.filter_query.as_str())
    }

    /// Return the current session ID of the filter widget, if active.
    pub fn session_id(&self) -> Option<egui::Id> {
        self.inner_state.as_ref().map(|state| state.session_id)
    }

    /// Return a filter matcher for the current query.
    pub fn filter(&self) -> FilterMatcher {
        FilterMatcher::new(self.query())
    }

    /// Display the filter widget.
    ///
    /// Note: this uses [`egui::Ui::available_width`] to determine the location of the right-aligned
    /// search button, as usual for [`list_item::ListItem`]-based widgets.
    pub fn ui(&mut self, ui: &mut egui::Ui, section_title: impl Into<egui::WidgetText>) {
        let mut button_clicked = false;

        let icon = if self.inner_state.is_none() {
            &crate::icons::SEARCH
        } else {
            &crate::icons::CLOSE
        };

        // precompute the title layout such that we know the size we need for the list item content
        let section_title = section_title.into();
        let galley = section_title.into_galley(
            ui,
            Some(egui::TextWrapMode::Extend),
            f32::INFINITY,
            egui::FontSelection::default(),
        );
        let text_width = galley.size().x;

        list_item::list_item_scope(ui, ui.next_auto_id(), |ui| {
            ui.list_item()
                .interactive(false)
                .with_height(30.0)
                .show_flat(
                    ui,
                    list_item::CustomContent::new(|ui, _| {
                        if let Some(inner_state) = self.inner_state.as_mut() {
                            // we add additional spacing for aesthetic reasons (active text edits have a
                            // fat border)
                            ui.spacing_mut().text_edit_width =
                                (ui.max_rect().width() - 10.0).at_least(0.0);

                            let response =
                                egui::TextEdit::singleline(&mut inner_state.filter_query)
                                    .lock_focus(true)
                                    .ui(ui);

                            if self.request_focus {
                                self.request_focus = false;
                                response.request_focus();
                            } else if response.lost_focus() {
                                self.inner_state = None;
                            }
                        } else {
                            ui.label(galley);
                        }
                    })
                    .with_content_width(text_width)
                    .action_button(icon, || {
                        button_clicked = true;
                    }),
                );
        });

        // defer button handling because we can't mutably borrow `self` in both closures above
        if button_clicked {
            if self.inner_state.is_none() {
                self.inner_state = Some(InnerState::default());
                self.request_focus = true;
            } else {
                self.inner_state = None;
            }
        }
    }
}

// --

/// Full-text, case-insensitive matcher.
pub struct FilterMatcher {
    /// The lowercase version of the query string.
    ///
    /// If this is `None` or `Some("")`, the matcher will accept any input.
    lowercase_query: Option<String>,
}

impl FilterMatcher {
    fn new(query: Option<&str>) -> Self {
        Self {
            lowercase_query: query.map(|s| s.to_lowercase()),
        }
    }

    /// Is the filter set to match everything?
    ///
    /// Can be used by client code to short-circuit more expansive matching logic.
    pub fn matches_everything(&self) -> bool {
        self.lowercase_query.is_none()
    }

    /// Is the filter set to match nothing?
    ///
    /// Can be used by client code to short-circuit more expansive matching logic.
    pub fn matches_nothing(&self) -> bool {
        self.lowercase_query
            .as_ref()
            .is_some_and(|query| query.is_empty())
    }

    /// Does the given text match the filter?
    pub fn matches(&self, text: &str) -> bool {
        match self.lowercase_query.as_deref() {
            None => true,
            Some("") => false,
            Some(query) => text.to_lowercase().contains(query),
        }
    }

    /// Find all matches in the given text.
    ///
    /// Can be used to format the text, e.g. with [`format_matching_text`].
    pub fn find_matches(&self, text: &str) -> Option<impl Iterator<Item = (usize, usize)> + '_> {
        let query = match self.lowercase_query.as_deref() {
            None => {
                return Some(Either::Left(std::iter::empty()));
            }
            Some("") => {
                return None;
            }
            Some(query) => query,
        };

        let mut start = 0;
        let lower_case_text = text.to_lowercase();
        let query_len = query.len();

        if !lower_case_text.contains(query) {
            return None;
        }

        Some(Either::Right(std::iter::from_fn(move || {
            let index = lower_case_text[start..].find(query)?;
            let start_index = start + index;
            start = start_index + query_len;
            Some((start_index, start_index + query_len))
        })))
    }

    pub fn matches_formatted(&self, ctx: &egui::Context, text: &str) -> Option<egui::WidgetText> {
        self.find_matches(text)
            .map(|match_iter| format_matching_text(ctx, text, match_iter))
    }
}

/// Given a list of highlight sections defined by start/end indices and a string, produce a properly
/// highlighted [`egui::WidgetText`].
pub fn format_matching_text(
    ctx: &egui::Context,
    text: &str,
    match_iter: impl Iterator<Item = (usize, usize)>,
) -> egui::WidgetText {
    let mut current = 0;
    let mut job = egui::text::LayoutJob::default();

    for (start_idx, end_idx) in match_iter {
        if current < start_idx {
            job.append(
                &text[current..start_idx],
                0.0,
                egui::TextFormat {
                    font_id: egui::TextStyle::Body.resolve(&ctx.style()),
                    color: Color32::PLACEHOLDER,
                    ..Default::default()
                },
            );
        }

        job.append(
            &text[start_idx..end_idx],
            0.0,
            egui::TextFormat {
                font_id: egui::TextStyle::Body.resolve(&ctx.style()),
                color: Color32::PLACEHOLDER,
                background: ctx.style().visuals.selection.bg_fill,
                ..Default::default()
            },
        );

        current = end_idx;
    }

    if current < text.len() {
        job.append(
            &text[current..],
            0.0,
            egui::TextFormat {
                font_id: egui::TextStyle::Body.resolve(&ctx.style()),
                color: Color32::PLACEHOLDER,
                ..Default::default()
            },
        );
    }

    job.into()
}
