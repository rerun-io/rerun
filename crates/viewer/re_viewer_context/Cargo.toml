[package]
authors.workspace = true
description = "Rerun viewer state that is shared with the viewer's code components."
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "re_viewer_context"
publish = true
readme = "README.md"
repository.workspace = true
rust-version.workspace = true
version.workspace = true
include = ["../../LICENSE-APACHE", "../../LICENSE-MIT", "**/*.rs", "Cargo.toml"]

[lints]
workspace = true

[package.metadata.docs.rs]
all-features = true

[dependencies]
re_chunk_store.workspace = true
re_chunk.workspace = true
re_data_source.workspace = true
re_entity_db = { workspace = true, features = ["serde"] }
re_format.workspace = true
re_log_types.workspace = true
re_log.workspace = true
re_math.workspace = true
re_query.workspace = true
re_renderer = { workspace = true, features = ["serde"] }
re_smart_channel.workspace = true
re_string_interner.workspace = true
re_tracing.workspace = true
re_types = { workspace = true, features = ["ecolor", "glam", "image"] }
re_types_core.workspace = true
re_ui.workspace = true
re_video = { workspace = true, features = ["serde"] }

ahash.workspace = true
anyhow.workspace = true
bit-vec.workspace = true
bitflags.workspace = true
bytemuck.workspace = true
directories.workspace = true
egui_extras.workspace = true
egui_tiles.workspace = true
egui-wgpu.workspace = true
egui.workspace = true
emath.workspace = true
glam = { workspace = true, features = ["serde"] }
half.workspace = true
image = { workspace = true, features = ["jpeg", "png"] }
indexmap = { workspace = true, features = ["std", "serde"] }
itertools.workspace = true
linked-hash-map.workspace = true
ndarray.workspace = true
nohash-hasher.workspace = true
once_cell.workspace = true
parking_lot = { workspace = true, features = ["serde"] }
rfd.workspace = true
serde.workspace = true
slotmap.workspace = true
smallvec.workspace = true
strum_macros.workspace = true
thiserror.workspace = true
uuid = { workspace = true, features = ["serde", "v4", "js"] }
wgpu.workspace = true

# Native dependencies:
[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
arboard = { workspace = true, default-features = false, features = [
  "image-data",
] }

# Web dependencies:
[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm-bindgen-futures.workspace = true
