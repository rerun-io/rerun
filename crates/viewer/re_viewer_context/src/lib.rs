//! Rerun Viewer context
//!
//! This crate contains data structures that are shared with most modules of the viewer.

mod annotations;
mod app_options;
mod blueprint_helpers;
mod blueprint_id;
mod cache;
mod collapsed_id;
mod command_sender;
mod component_fallbacks;
mod component_ui_registry;
mod contents;
mod file_dialog;
mod image_info;
mod item;
mod maybe_mut_ref;
mod query_context;
mod query_range;
mod selection_history;
mod selection_state;
mod space_view;
mod store_context;
pub mod store_hub;
mod tensor;
pub mod test_context; //TODO(ab): this should be behind #[cfg(test)], but then ` cargo clippy --all-targets` fails
mod time_control;
mod time_drag_value;
mod typed_entity_collections;
mod utils;
mod viewer_context;

// TODO(andreas): Move to its own crate?
pub mod gpu_bridge;

pub use annotations::{
    AnnotationMap, Annotations, ResolvedAnnotationInfo, ResolvedAnnotationInfos,
};
pub use app_options::AppOptions;
pub use blueprint_helpers::{blueprint_timeline, blueprint_timepoint_for_writes};
pub use blueprint_id::{BlueprintId, BlueprintIdRegistry, ContainerId, SpaceViewId};
pub use cache::{Cache, Caches, ImageDecodeCache, ImageStatsCache, TensorStatsCache, VideoCache};
pub use collapsed_id::{CollapseItem, CollapseScope, CollapsedId};
pub use command_sender::{
    command_channel, CommandReceiver, CommandSender, SystemCommand, SystemCommandSender,
};
pub use component_fallbacks::{
    ComponentFallbackError, ComponentFallbackProvider, ComponentFallbackProviderResult,
    TypedComponentFallbackProvider,
};
pub use component_ui_registry::{ComponentUiRegistry, ComponentUiTypes, UiLayout};
pub use contents::{blueprint_id_to_tile_id, Contents, ContentsName};
pub use image_info::ImageInfo;
pub use item::Item;
pub use maybe_mut_ref::MaybeMutRef;
pub use query_context::{
    DataQueryResult, DataResultHandle, DataResultNode, DataResultTree, QueryContext,
};
pub use query_range::QueryRange;
pub use selection_history::SelectionHistory;
pub use selection_state::{
    ApplicationSelectionState, HoverHighlight, InteractionHighlight, ItemCollection,
    ItemSpaceContext, SelectionHighlight,
};
pub use space_view::{
    DataResult, IdentifiedViewSystem, OptionalSpaceViewEntityHighlight, OverridePath,
    PerSystemDataResults, PerSystemEntities, PropertyOverrides, RecommendedSpaceView,
    SmallVisualizerSet, SpaceViewClass, SpaceViewClassExt, SpaceViewClassLayoutPriority,
    SpaceViewClassRegistry, SpaceViewClassRegistryError, SpaceViewEntityHighlight,
    SpaceViewHighlights, SpaceViewOutlineMasks, SpaceViewSpawnHeuristics, SpaceViewState,
    SpaceViewStateExt, SpaceViewSystemExecutionError, SpaceViewSystemRegistrator,
    SystemExecutionOutput, ViewContext, ViewContextCollection, ViewContextSystem, ViewQuery,
    ViewStates, ViewSystemIdentifier, VisualizableFilterContext,
    VisualizerAdditionalApplicabilityFilter, VisualizerCollection, VisualizerQueryInfo,
    VisualizerSystem,
};
pub use store_context::StoreContext;
pub use store_hub::StoreHub;
pub use tensor::{ImageStats, TensorStats};
pub use time_control::{Looping, PlayState, TimeControl, TimeView};
pub use time_drag_value::TimeDragValue;
pub use typed_entity_collections::{
    ApplicableEntities, IndicatedEntities, PerVisualizer, VisualizableEntities,
};
pub use utils::{auto_color_egui, auto_color_for_entity_path, level_to_rich_text};
pub use viewer_context::{RecordingConfig, ViewerContext};

#[cfg(not(target_arch = "wasm32"))]
mod clipboard;

#[cfg(not(target_arch = "wasm32"))]
pub use clipboard::Clipboard;

pub mod external {
    pub use nohash_hasher;
    pub use {re_chunk_store, re_entity_db, re_log_types, re_query, re_ui};
}

// ---------------------------------------------------------------------------

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum NeedsRepaint {
    Yes,
    No,
}

// ---

/// Determines the icon to use for a given container kind.
#[inline]
pub fn icon_for_container_kind(kind: &egui_tiles::ContainerKind) -> &'static re_ui::Icon {
    match kind {
        egui_tiles::ContainerKind::Tabs => &re_ui::icons::CONTAINER_TABS,
        egui_tiles::ContainerKind::Horizontal => &re_ui::icons::CONTAINER_HORIZONTAL,
        egui_tiles::ContainerKind::Vertical => &re_ui::icons::CONTAINER_VERTICAL,
        egui_tiles::ContainerKind::Grid => &re_ui::icons::CONTAINER_GRID,
    }
}

/// The style to use for displaying this space view name in the UI.
pub fn contents_name_style(name: &ContentsName) -> re_ui::LabelStyle {
    match name {
        ContentsName::Named(_) => re_ui::LabelStyle::Normal,
        ContentsName::Placeholder(_) => re_ui::LabelStyle::Unnamed,
    }
}
