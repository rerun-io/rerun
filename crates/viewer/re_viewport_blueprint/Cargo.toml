[package]
authors.workspace = true
description = "The data model describing the layout of the viewport."
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "re_viewport_blueprint"
publish = true
readme = "README.md"
repository.workspace = true
rust-version.workspace = true
version.workspace = true
include = ["../../LICENSE-APACHE", "../../LICENSE-MIT", "**/*.rs", "Cargo.toml"]

[lints]
workspace = true

[package.metadata.docs.rs]
all-features = true

[dependencies]
re_chunk.workspace = true
re_chunk_store.workspace = true
re_entity_db.workspace = true
re_log.workspace = true
re_log_types.workspace = true
re_tracing.workspace = true
re_types_blueprint.workspace = true
re_types_core.workspace = true
re_types.workspace = true
re_ui.workspace = true
re_viewer_context.workspace = true

ahash.workspace = true
egui_tiles.workspace = true
egui.workspace = true
itertools.workspace = true
nohash-hasher.workspace = true
once_cell.workspace = true
parking_lot.workspace = true
slotmap.workspace = true
smallvec.workspace = true
thiserror.workspace = true
