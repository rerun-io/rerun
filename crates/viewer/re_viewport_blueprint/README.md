# re_viewport_blueprint

Part of the [`rerun`](https://github.com/rerun-io/rerun) family of crates.

[![Latest version](https://img.shields.io/crates/v/re_viewport_blueprint.svg)](https://crates.io/crates/re_viewport_blueprint?speculative-link)
[![Documentation](https://docs.rs/re_viewport_blueprint/badge.svg)](https://docs.rs/re_viewport_blueprint?speculative-link)
![MIT](https://img.shields.io/badge/license-MIT-blue.svg)
![Apache](https://img.shields.io/badge/license-Apache-blue.svg)

The data model describing the layout of the viewport.
