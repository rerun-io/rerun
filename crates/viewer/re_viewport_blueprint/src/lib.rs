//! Rerun Viewport Blueprint
//!
//! This crate provides blueprint (i.e. description) for how to render the viewport.

mod auto_layout;
mod container;
mod entity_add_info;
#[cfg(feature = "testing")]
pub mod test_context_ext;
pub mod ui;
mod view;
mod view_contents;
mod view_properties;
mod viewport_blueprint;
mod viewport_command;

pub use container::ContainerBlueprint;
pub use entity_add_info::{create_entity_add_info, CanAddToView, EntityAddInfo};
use re_viewer_context::ViewerContext;
pub use view::ViewBlueprint;
pub use view_contents::ViewContents;
pub use view_properties::{entity_path_for_view_property, ViewProperty, ViewPropertyQueryError};
pub use viewport_blueprint::{tree_simplification_options, ViewportBlueprint};
pub use viewport_command::ViewportCommand;

/// The entity path of the viewport blueprint in the blueprint store.
pub const VIEWPORT_PATH: &str = "viewport";

/// Converts a [`re_types::blueprint::components::ContainerKind`] into a [`egui_tiles::ContainerKind`].
///
/// Does not implement the `From`/`To` traits because we don't want `re_types` to depend
/// on `egui`, and we cannot do it from here because of orphan rules.
#[inline]
pub fn container_kind_to_egui(
    kind: re_types::blueprint::components::ContainerKind,
) -> egui_tiles::ContainerKind {
    use re_types::blueprint::components::ContainerKind;
    match kind {
        ContainerKind::Tabs => egui_tiles::ContainerKind::Tabs,
        ContainerKind::Horizontal => egui_tiles::ContainerKind::Horizontal,
        ContainerKind::Vertical => egui_tiles::ContainerKind::Vertical,
        ContainerKind::Grid => egui_tiles::ContainerKind::Grid,
    }
}

/// Converts a [`egui_tiles::ContainerKind`] into a [`re_types::blueprint::components::ContainerKind`].
///
/// Does not implement the `From`/`To` traits because we don't want `re_types` to depend
/// on `egui`, and we cannot do it from here because of orphan rules.
#[inline]
pub fn container_kind_from_egui(
    kind: egui_tiles::ContainerKind,
) -> re_types::blueprint::components::ContainerKind {
    use re_types::blueprint::components::ContainerKind;
    match kind {
        egui_tiles::ContainerKind::Tabs => ContainerKind::Tabs,
        egui_tiles::ContainerKind::Horizontal => ContainerKind::Horizontal,
        egui_tiles::ContainerKind::Vertical => ContainerKind::Vertical,
        egui_tiles::ContainerKind::Grid => ContainerKind::Grid,
    }
}

/// List out all views we generate by default for the available data.
///
/// TODO(andreas): This is transitional. We want to pass on the view spawn heuristics
/// directly and make more high level decisions with it.
pub fn default_created_views(ctx: &ViewerContext<'_>) -> Vec<ViewBlueprint> {
    re_tracing::profile_function!();

    ctx.view_class_registry
        .iter_registry()
        .flat_map(|entry| {
            let spawn_heuristics = entry.class.spawn_heuristics(ctx);
            spawn_heuristics
                .into_vec()
                .into_iter()
                .map(|recommendation| ViewBlueprint::new(entry.identifier, recommendation))
        })
        .collect()
}
