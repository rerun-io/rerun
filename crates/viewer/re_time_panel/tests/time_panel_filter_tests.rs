#![cfg(feature = "testing")]

use egui::Vec2;

use re_chunk_store::external::re_chunk::ChunkBuilder;
use re_chunk_store::{LatestAtQuery, RowId};
use re_log_types::TimePoint;
use re_time_panel::StreamsTreeData;
use re_time_panel::{TimePanel, TimePanelSource};
use re_types::archetypes::Points3D;
use re_ui::filter_widget::FilterState;
use re_viewer_context::{blueprint_timeline, test_context::TestContext};
use re_viewport_blueprint::ViewportBlueprint;

fn filter_queries() -> impl Iterator<Item = Option<&'static str>> {
    [
        None,
        Some("t"),
        Some("void"),
        Some("path"),
        Some("ath t"),
        Some("ath left"),
        Some("to/the"),
        Some("/to/the"),
        Some("to/the/"),
        Some("/to/the/"),
        Some("to/the oid"),
        Some("/path/to /rig"),
    ]
    .into_iter()
}

#[test]
pub fn test_various_filter_ui_snapshot() {
    TimePanel::ensure_registered_subscribers();

    for filter_query in filter_queries() {
        let test_context = prepare_test_context();

        let mut time_panel = TimePanel::default();
        if let Some(query) = filter_query {
            time_panel.activate_filter(query);
        }

        run_time_panel_and_save_snapshot(
            test_context,
            time_panel,
            &format!(
                "various_filters-{}",
                filter_query
                    .map(|s| s.replace(' ', ",").replace('/', "_"))
                    .unwrap_or("none".to_owned())
            ),
        );
    }
}

#[test]
pub fn test_various_filter_insta_snapshot() {
    for filter_query in filter_queries() {
        let test_context = prepare_test_context();

        let streams_tree_data = test_context.run_once_in_egui_central_panel(|viewer_ctx, _| {
            let mut filter_state = FilterState::default();

            if let Some(filter_query) = filter_query {
                filter_state.activate(filter_query);
            }

            StreamsTreeData::from_source_and_filter(
                viewer_ctx,
                TimePanelSource::Recording,
                &filter_state.filter(),
            )
        });

        let snapshot_name = format!(
            "various_filters-{}",
            filter_query
                .map(|s| s.replace(' ', ",").replace('/', "_"))
                .unwrap_or("none".to_owned())
        );

        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.bind(|| {
            insta::assert_yaml_snapshot!(snapshot_name, streams_tree_data);
        });
    }
}

fn prepare_test_context() -> TestContext {
    let mut test_context = TestContext::default();

    test_context.log_entity("/path/to/left".into(), add_point_to_chunk_builder);
    test_context.log_entity("/path/to/right".into(), add_point_to_chunk_builder);
    test_context.log_entity("/path/to/the/void".into(), add_point_to_chunk_builder);
    test_context.log_entity("/path/onto/their/coils".into(), add_point_to_chunk_builder);
    test_context.log_entity("/center/way".into(), add_point_to_chunk_builder);

    // also populate some "intermediate" entities so we see components
    test_context.log_entity("/path".into(), add_point_to_chunk_builder);
    test_context.log_entity("/path/to".into(), add_point_to_chunk_builder);

    test_context
}

fn add_point_to_chunk_builder(builder: ChunkBuilder) -> ChunkBuilder {
    // log as static to minimize "noise" in the snapshot
    builder.with_archetype(
        RowId::new(),
        TimePoint::default(),
        &Points3D::new([[0.0, 0.0, 0.0]]),
    )
}

fn run_time_panel_and_save_snapshot(
    mut test_context: TestContext,
    mut time_panel: TimePanel,
    snapshot_name: &str,
) {
    let mut harness = test_context
        .setup_kittest_for_rendering()
        .with_size(Vec2::new(700.0, 700.0))
        .build_ui(|ui| {
            test_context.run(&ui.ctx().clone(), |viewer_ctx| {
                let blueprint = ViewportBlueprint::try_from_db(
                    viewer_ctx.store_context.blueprint,
                    &LatestAtQuery::latest(blueprint_timeline()),
                );

                let mut time_ctrl = viewer_ctx.rec_cfg.time_ctrl.read().clone();

                time_panel.show_expanded_with_header(
                    viewer_ctx,
                    &blueprint,
                    viewer_ctx.recording(),
                    &mut time_ctrl,
                    ui,
                );

                *viewer_ctx.rec_cfg.time_ctrl.write() = time_ctrl;
            });

            test_context.handle_system_commands();
        });

    harness.run();
    harness.snapshot(snapshot_name);
}
