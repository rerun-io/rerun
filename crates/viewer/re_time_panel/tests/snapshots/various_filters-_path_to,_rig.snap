---
source: crates/viewer/re_time_panel/tests/time_panel_filter_tests.rs
expression: streams_tree_data
---
children:
  - entity_path: []
    label: /
    highlight_sections: []
    default_open: true
    children:
      - entity_path:
          - path
        label: path/
        highlight_sections:
          - start: 0
            end: 4
        default_open: true
        children:
          - entity_path:
              - path
              - to
            label: to/
            highlight_sections:
              - start: 0
                end: 2
            default_open: true
            children:
              - entity_path:
                  - path
                  - to
                  - right
                label: right
                highlight_sections:
                  - start: 0
                    end: 3
                default_open: false
                children: []
